### PR TITLE
SWATCH-1373 Support event lookup by both UOM and metricID

### PR DESCRIPTION
**Pull request by kartikshahc**
Jira issue: [SWATCH-1373](https://issues.redhat.com/browse/SWATCH-1373)

### Setup

1. To run locally prometheus by pointing to stage follow the steps in: [Connecting  Observatorium to fetch Prometheus ](https://docs.engineering.redhat.com/display/ENT/Connecting+to+Observatorium+to+Fetch+Prometheus+Metrics)
- Copy CLIENT_SECRET and paste it in a file `token-refresher`
- Run: `podman run --name=telemeter-stage --rm -ti -p 8082:8080     quay.io/observatorium/token-refresher:master-2021-02-05-5da9663     --oidc.client-secret=$(cat ~/Downloads/token-refresher)     --oidc.client-id=observatorium-subwatch-staging     --oidc.issuer-url=https://sso.redhat.com/auth/realms/redhat-external     --url=https://observatorium.api.stage.openshift.com`
- Then in separate terminal pointing to the prom_url in the doc run: `RHSM_SUBSCRIPTIONS_ENABLE_SYNCHRONOUS_OPERATIONS=true DEV_MODE=true PROM_URL=http://localhost:8082/api/metrics/v1/telemeter/api/v1 ./gradlew clean :bootRun`

2. Insert few records in Events table:
```
INSERT INTO public.events (id, account_number, timestamp, data, event_type, event_source, instance_id, org_id) VALUES ('2b16a0a1-8278-488a-a8cd-67178a07a013', null, '2023-07-05 15:54:00.000000 +00:00', '{"sla": "Premium", "role": "BASILISK", "org_id": "16785082", "event_id": "2b16a0a1-8278-488a-a8cd-67178a07a013", "timestamp": "2023-06-07T08:00:00Z", "event_type": "snapshot_redhat.com:BASILISK:transfer_gb", "expiration": "2023-06-07T09:00:00Z", "instance_id": "RSsUTrGhOO", "display_name": "automation_BASILISK_cluster_RSsUTrGhOO", "event_source": "prometheus", "measurements": [{"uom": "Transfer-gibibytes", "value": 1.0}], "service_type": "BASILISK Instance"}', 'snapshot_redhat.com:BASILISK:transfer_gb', 'prometheus', 'RSsUTrGhOO', '16785082');
INSERT INTO public.events (id, account_number, timestamp, data, event_type, event_source, instance_id, org_id) VALUES ('112a10c5-1374-4c95-964c-9defb5e4c6f6', null, '2023-07-05 15:54:00.000000 +00:00', '{"sla": "Premium", "role": "rhosak", "org_id": "16785082", "event_id": "112a10c5-1374-4c95-964c-9defb5e4c6f6", "timestamp": "2023-07-05T15:54:00Z", "event_type": "snapshot_rhosak_transfer_gibibytes", "expiration": "2023-07-05T13:54:00Z", "instance_id": "56ad40fe-b9bd-49a5-8c6c-a5478ad98a2e", "display_name": "automation_rhosak_cluster_56ad40fe-b9bd-49a5-8c6c-a5478ad98a2e", "event_source": "prometheus", "measurements": [{"uom": "Transfer-gibibytes", "value": 1.0}], "service_type": "Kafka Cluster", "billing_provider": "red hat", "billing_account_id": "4accf275-7db2-4199-9cb6-e7650bd0aa45"}', 'snapshot_rhosak_transfer-gibibytes', 'prometheus', '56ad40fe-b9bd-49a5-8c6c-a5478ad98a2e', '16785082');
INSERT INTO public.events (id, account_number, timestamp, data, event_type, event_source, instance_id, org_id) VALUES ('112a10c5-1374-4c95-964c-9defb5e4c6f3', null, '2023-07-05 15:54:00.000000 +00:00', '{"sla": "Premium", "role": "rhosak", "org_id": "16785082", "event_id": "112a10c5-1374-4c95-964c-9defb5e4c6f3", "timestamp": "2023-07-05T15:54:00Z", "event_type": "snapshot_redhat.com:rhosak:transfer_gb", "expiration": "2023-06-07T06:00:00Z", "instance_id": "56ad40fe-b9bd-49a5-8c6c-a5478ad98a2e", "display_name": "automation_rhosak_cluster_56ad40fe-b9bd-49a5-8c6c-a5478ad98a2e", "event_source": "prometheus", "measurements": [{"uom": "Transfer-gibibytes", "value": 1.0}], "service_type": "Kafka Cluster", "billing_provider": "red hat", "billing_account_id": "4accf275-7db2-4199-9cb6-e7650bd0aa45"}', 'snapshot_redhat.com:rhosak:transfer_gb', 'prometheus', '56ad40fe-b9bd-49a5-8c6c-a5478ad98a2e', '16785082');
INSERT INTO public.events (id, account_number, timestamp, data, event_type, event_source, instance_id, org_id) VALUES ('a94e8108-32d4-407f-9a50-d50880b3e0e1', null, '2023-07-05 15:54:00.000000 +00:00', '{"sla": "Premium", "role": "rhosak", "org_id": "16785082", "event_id": "a94e8108-32d4-407f-9a50-d50880b3e0e1", "timestamp": "2023-07-05T15:54:00Z", "event_type": "snapshot_redhat.com:rhosak:transfer_gb", "expiration": "2023-06-07T06:00:00Z", "instance_id": "21a0c030-8706-4b46-9836-5ded5ca9a7cc", "display_name": "automation_rhosak_cluster_21a0c030-8706-4b46-9836-5ded5ca9a7cc", "event_source": "prometheus", "measurements": [{"uom": "Transfer-gibibytes", "value": 1.0}], "service_type": "Kafka Cluster", "billing_provider": "gcp", "billing_account_id": "4accf275-7db2-4199-9cb6-e7650bd0aa45"}', 'snapshot_redhat.com:rhosak:transfer_gb', 'prometheus', '21a0c030-8706-4b46-9836-5ded5ca9a7cc', '16785082');
INSERT INTO public.events (id, account_number, timestamp, data, event_type, event_source, instance_id, org_id) VALUES ('7a375596-484c-4736-9c1e-52362d965f17', null, '2023-07-05 15:54:00.000000 +00:00', '{"sla": "Premium", "role": "osd", "org_id": "16785082", "event_id": "7a375596-484c-4736-9c1e-52362d965f17", "timestamp": "2023-07-05T15:54:00Z", "event_type": "snapshot_redhat.com:openshift_dedicated:cluster_hour", "expiration": "2023-05-01T01:00:00Z", "instance_id": "93526e44-1296-4674-89ec-1213d44c5ae9", "display_name": "automation_osd_cluster_93526e44-1296-4674-89ec-1213d44c5ae9", "event_source": "prometheus", "measurements": [{"uom": "Instance-hours", "value": 1.0}], "service_type": "OpenShift Cluster"}', 'snapshot_redhat.com:openshift_dedicated:cluster_hour', 'prometheus', '93526e44-1296-4674-89ec-1213d44c5ae9', '16785082');
INSERT INTO public.events (id, account_number, timestamp, data, event_type, event_source, instance_id, org_id) VALUES ('7a375596-484c-4736-9c1e-52362d965f18', null, '2023-07-05 15:54:00.000000 +00:00', '{"sla": "Premium", "role": "osd", "org_id": "16785082", "event_id": "7a375596-484c-4736-9c1e-52362d965f18", "timestamp": "2023-07-05T15:54:00Z", "event_type": "snapshot_openshift-dedicated-metrics_instance-hours", "expiration": "2023-05-01T01:00:00Z", "instance_id": "93526e44-1296-4674-89ec-1213d44c5ae9", "display_name": "automation_osd_cluster_93526e44-1296-4674-89ec-1213d44c5ae9", "event_source": "prometheus", "measurements": [{"uom": "Instance-hours", "value": 1.0}], "service_type": "OpenShift Cluster"}', 'snapshot_openshift-dedicated-metrics_instance-hours', 'prometheus', '93526e44-1296-4674-89ec-1213d44c5ae9', '16785082');
INSERT INTO public.events (id, account_number, timestamp, data, event_type, event_source, instance_id, org_id) VALUES ('2b16a0a1-8278-488a-a8cd-67178a07a014', null, '2023-07-05 15:54:00.000000 +00:00', '{"sla": "Premium", "role": "BASILISK", "org_id": "16785082", "event_id": "2b16a0a1-8278-488a-a8cd-67178a07a014", "timestamp": "2023-07-05T15:54:00Z", "event_type": "snapshot_basilisk_transfer-gibibytes", "expiration": "2023-06-07T09:00:00Z", "instance_id": "RSsUTrGhOO", "display_name": "automation_BASILISK_cluster_RSsUTrGhOO", "event_source": "prometheus", "measurements": [{"uom": "Transfer-gibibytes", "value": 1.0}], "service_type": "BASILISK Instance"}', 'snapshot_basilisk_transfer-gibibytes', 'prometheus', 'RSsUTrGhOO', '16785082');

```

### Steps

1. `curl -X POST -H "x-rh-swatch-synchronous-request:true"  -H "x-rh-swatch-psk: placeholder" "http://localhost:8000/api/rhsm-subscriptions/v1/internal/metering/rhosak?orgId=16785082&rangeInMinutes=1200"`

### Verification

1. In debug mode you will see below logs:
```
2023-07-05 20:33:23,545 [thread=http-nio-8000-exec-1] [DEBUG] [org.hibernate.SQL] self- select eventrecor0_.id as id1_3_, eventrecor0_.account_number as account_2_3_, eventrecor0_.data as data3_3_, eventrecor0_.event_source as event_so4_3_, eventrecor0_.event_type as event_ty5_3_, eventrecor0_.instance_id as instance6_3_, eventrecor0_.org_id as org_id7_3_, eventrecor0_.timestamp as timestam8_3_ from events eventrecor0_ where eventrecor0_.org_id=? and eventrecor0_.event_source=? and (eventrecor0_.event_type in (? , ?)) and eventrecor0_.timestamp>=? and eventrecor0_.timestamp<? order by eventrecor0_.timestamp asc

2023-07-05 20:40:59,529 [thread=http-nio-8000-exec-8] [INFO ] [org.candlepin.subscriptions.metering.service.prometheus.PrometheusMeteringController] self- Deleting 3 stale metric events.
2023-07-05 20:40:59,530 [thread=http-nio-8000-exec-8] [DEBUG] [org.hibernate.SQL] self- delete from events where id=? or id=? or id=?

```
Without any errors it will delete the events that are: `snapshot_rhosak_transfer-gibibytes` and `snapshot_redhat.com:rhosak:transfer_gb`

### Steps

2. `curl -X POST -H "x-rh-swatch-synchronous-request:true"  -H "x-rh-swatch-psk: placeholder" "http://localhost:8000/api/rhsm-subscriptions/v1/internal/metering/OpenShift-dedicated-metrics?orgId=16785082&rangeInMinutes=1200"`

### Verification

2. Logs as below and it will delete the events with `snapshot_redhat.com:openshift_dedicated:cluster_hour` and `snapshot_openshift-dedicated-metrics_instance-hours`
```
2023-07-05 20:45:16,695 [thread=http-nio-8000-exec-9] [DEBUG] [org.hibernate.SQL] self- select eventrecor0_.id as id1_3_, eventrecor0_.account_number as account_2_3_, eventrecor0_.data as data3_3_, eventrecor0_.event_source as event_so4_3_, eventrecor0_.event_type as event_ty5_3_, eventrecor0_.instance_id as instance6_3_, eventrecor0_.org_id as org_id7_3_, eventrecor0_.timestamp as timestam8_3_ from events eventrecor0_ where eventrecor0_.org_id=? and eventrecor0_.event_source=? and (eventrecor0_.event_type in (? , ?)) and eventrecor0_.timestamp>=? and eventrecor0_.timestamp<? order by eventrecor0_.timestamp asc
2023-07-05 20:45:16,697 [thread=http-nio-8000-exec-9] [INFO ] [org.candlepin.subscriptions.metering.service.prometheus.PrometheusMeteringController] self- Persisted 0 events for OpenShift-dedicated-metrics Instance-hours metrics.
2023-07-05 20:45:16,697 [thread=http-nio-8000-exec-9] [INFO ] [org.candlepin.subscriptions.metering.service.prometheus.PrometheusMeteringController] self- Deleting 2 stale metric events.
2023-07-05 20:45:16,698 [thread=http-nio-8000-exec-9] [DEBUG] [org.hibernate.SQL] self- delete from events where id=? or id=?

```

### Steps

3. `curl -X POST -H "x-rh-swatch-synchronous-request:true"  -H "x-rh-swatch-psk: placeholder" "http://localhost:8000/api/rhsm-subscriptions/v1/internal/metering/BASILISK?orgId=16785082&rangeInMinutes=1200"`

### Verification

3. It will delete the events with `snapshot_redhat.com:BASILISK:transfer_gb` and `snapshot_basilisk_transfer-gibibytes`

To test on EE:
`curl -X POST -H 'x-rh-swatch-synchronous-request:true' -H 'x-rh-swatch-psk:dummy' -H 'Origin: https://cloud.redhat.com/' "http://localhost:8000/api/rhsm-subscriptions/v1/internal/metering/rhosak?orgId=16785082&rangeInMinutes=1200"`

### CODE REVIEW

**Comment by ssidhaye** [comment]
```
curl -X POST -H 'x-rh-swatch-synchronous-request:true' -H 'x-rh-swatch-psk:dummy' -H 'Origin: https://cloud.redhat.com' "http://localhost:8000/api/rhsm-subscriptions/v1/internal/metering/rhosak?orgId=3340851&rangeInMinutes=1200"
2023-07-06 15:18:12,999 [thread=http-nio-8000-exec-1] [INFO ] [org.candlepin.subscriptions.metering.service.prometheus.PrometheusService] self- Fetching metrics from prometheus: 2023-07-05T20:00Z -> 2023-07-06T15:00Z [Step: 3600]
2023-07-06 15:18:13,086 [thread=http-nio-8000-exec-1] [INFO ] [org.candlepin.subscriptions.metering.service.prometheus.PrometheusMeteringController] self- Persisted 0 events for rhosak Transfer-gibibytes metrics.
2023-07-06 15:18:13,086 [thread=http-nio-8000-exec-1] [INFO ] [org.candlepin.subscriptions.metering.service.prometheus.PrometheusMeteringController] self- Deleting 1 stale metric events.

curl -X POST -H 'x-rh-swatch-synchronous-request:true' -H 'x-rh-swatch-psk:dummy' -H 'Origin: https://cloud.redhat.com' "http://localhost:8000/api/rhsm-subscriptions/v1/internal/metering/OpenShift-dedicated-metrics?orgId=3340851&rangeInMinutes=1200"

2023-07-06 15:23:23,134 [thread=http-nio-8000-exec-6] [INFO ] [org.candlepin.subscriptions.metering.service.prometheus.PrometheusMeteringController] self- Collecting metrics for orgId=3340851: OpenShift-dedicated-metrics Instance-hours
2023-07-06 15:23:23,135 [thread=http-nio-8000-exec-6] [INFO ] [org.candlepin.subscriptions.metering.service.prometheus.PrometheusService] self- Fetching metrics from prometheus: 2023-07-05T20:00Z -> 2023-07-06T15:00Z [Step: 3600]
2023-07-06 15:23:23,145 [thread=http-nio-8000-exec-6] [INFO ] [org.candlepin.subscriptions.metering.service.prometheus.PrometheusMeteringController] self- Persisted 0 events for OpenShift-dedicated-metrics Instance-hours metrics.
2023-07-06 15:23:23,145 [thread=http-nio-8000-exec-6] [INFO ] [org.candlepin.subscriptions.metering.service.prometheus.PrometheusMeteringController] self- Deleting 1 stale metric events.

curl -X POST -H 'x-rh-swatch-synchronous-request:true' -H 'x-rh-swatch-psk:dummy' -H 'Origin: https://cloud.redhat.com' "http://localhost:8000/api/rhsm-subscriptions/v1/internal/metering/BASILISK?orgId=3340851&rangeInMinutes=1200"

023-07-06 15:25:26,574 [thread=http-nio-8000-exec-8] [INFO ] [org.candlepin.subscriptions.metering.service.prometheus.PrometheusService] self- Fetching metrics from prometheus: 2023-07-05T20:00Z -> 2023-07-06T15:00Z [Step: 3600]
2023-07-06 15:25:26,582 [thread=http-nio-8000-exec-8] [INFO ] [org.candlepin.subscriptions.metering.service.prometheus.PrometheusMeteringController] self- Persisted 0 events for BASILISK Transfer-gibibytes metrics.
2023-07-06 15:25:26,582 [thread=http-nio-8000-exec-8] [INFO ] [org.candlepin.subscriptions.metering.service.prometheus.PrometheusMeteringController] self- Deleting 2 stale metric events.
```


Records are also removed from the DB once curl is called indicating the functionality is working as expected

File: swatch-core/src/main/java/org/candlepin/subscriptions/db/EventRecordRepository.java [hunk 1]
```diff
@@ -73,10 +74,10 @@ Stream<EventRecord> findByOrgIdAndTimestampGreaterThanEqualAndTimestampLessThanO
    * @return Stream of EventRecords
    */
   Stream<EventRecord>
-      findByOrgIdAndEventSourceAndEventTypeAndTimestampGreaterThanEqualAndTimestampLessThanOrderByTimestamp(
+      findByOrgIdAndEventSourceAndEventTypeIsInAndTimestampGreaterThanEqualAndTimestampLessThanOrderByTimestamp(
```
Comment (awood) [inline swatch-core/src/main/java/org/candlepin/subscriptions/db/EventRecordRepository.java:77]: Can we alias this method?  That name is just too long to be able to easily comprehend.  We can just do something like 

```java
interface UserRepository extends JpaRepository<User, Long> {
  // don't use that crazy long method! use getByEmail instead
  User findFirstByEmailContainsIgnoreCaseAndField1NotNullAndField2NotNullAndField3NotNullAndField4NotNullAndField5NotNullAndField6NotNull(final String email);

  default User getByEmail(final String email) {
    return findFirstByEmailContainsIgnoreCaseAndField1NotNullAndField2NotNullAndField3NotNullAndField4NotNullAndField5NotNullAndField6NotNull(email);
  }
}
```


(from https://stackoverflow.com/a/53315665/6124862)

File: src/main/java/org/candlepin/subscriptions/event/EventController.java [hunk 2]
```diff
@@ -72,10 +72,10 @@ public Stream<Event> fetchEventsInTimeRangeByServiceType(
   public Map<EventKey, Event> mapEventsInTimeRange(
       String orgId,
       String eventSource,
-      String eventType,
+      List<String> eventType,
```
Comment (awood) [inline src/main/java/org/candlepin/subscriptions/event/EventController.java:75]: Should we make this a `Set`?  Seems like having duplicates in `eventType` wouldn't make any sense so we shouldn't even allow it in the first place.

File: src/main/java/org/candlepin/subscriptions/metering/MeteringEventFactory.java [hunk 3]
```diff
@@ -131,12 +133,31 @@ public static void updateMetricEvent(
         .withRole(getRole(role, accountNumber, instanceId));
   }
 
-  public static String getEventType(String metricId) {
+  // SWATCH-1374 Remove or update this method
+  public static String getOldEventType(String metricId) {
```
Comment (awood) [inline src/main/java/org/candlepin/subscriptions/metering/MeteringEventFactory.java:137]: We should just keep the same method name but mark this as `Deprecated` show that we'll get compiler and IDE warnings.

**Review by awood — CHANGES_REQUESTED** [review]
I'm not seeing anything getting deleted after following the instructions in the PR.  Ping me and we can walk through it and see what's gone awry.

**Comment by awood** [comment]
> I'm not seeing anything getting deleted after following the instructions in the PR. Ping me and we can walk through it and see what's gone awry.

Looks like @ssidhaye has it replicated so that is sufficient for me.  Let's just address those minor things I brought up in the other comments.

File: swatch-core/src/main/java/org/candlepin/subscriptions/db/EventRecordRepository.java (hunk 1 above)
Comment (kartikshahc) [reply swatch-core/src/main/java/org/candlepin/subscriptions/db/EventRecordRepository.java:77]: I have wrapped the code as shown in example.

File: src/main/java/org/candlepin/subscriptions/event/EventController.java (hunk 2 above)
Comment (kartikshahc) [reply src/main/java/org/candlepin/subscriptions/event/EventController.java]: Done

**Review by awood — APPROVED** [review]

**Comment by kartikshahc** [comment]
Merging the code since got acked from dev and QE

**Pull request merged**